# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # Create source_id_text list
    source_id_text = []
    for i, sentence in enumerate(source_text.split('\n')):
        source_id_text.append([source_vocab_to_int.get(word, source_vocab_to_int['<UNK>']) for word in sentence.split()])
                            
    # Create target_id_text list
    target_id_text = []
    for i, sentence in enumerate(target_text.split('\n')):
        target_id_text.append([target_vocab_to_int.get(word, target_vocab_to_int['<UNK>']) for word in sentence.split()])
        # Add <EOS> id to last position of target_id_text
        target_id_text[i].append(target_vocab_to_int['<EOS>'])

    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
print('TensorFlow Version: {}'.format(tf.__version__))
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    input_ = tf.placeholder(tf.int32, [None, None], name="input")
    targets = tf.placeholder(tf.int32, [None, None], name="targets")
    lr = tf.placeholder(tf.float32, name="learning_rate")
    keep_prob = tf.placeholder(tf.float32, name="keep_prob")
    target_sequence_length = tf.placeholder(tf.int32, [None], name="target_sequence_length")
    max_target_len = tf.reduce_max(target_sequence_length, name="max_target_len")
    source_sequence_length = tf.placeholder(tf.int32, [None], name="source_sequence_length")
    return input_, targets, lr, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # Remove last element from each sequence of the batch
    ending = tf.strided_slice(target_data, [0,0], [batch_size, -1], [1,1])
    # Add <GO> id to the end of each sequence in the batch
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    
    # Embed input into encoded input matrix
    encoder_embedded_input = tf.contrib.layers.embed_sequence(rnn_inputs, 
                                                              source_vocab_size, 
                                                              encoding_embedding_size)
    
    # Function to create a single LSTM Cell witn rnn_size
    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1,0.1, seed=2))
        enc_cell = tf.contrib.rnn.DropoutWrapper(enc_cell, keep_prob)
        return enc_cell
    
    # Stack LSTM cells
    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    # Pass LSTM cells and embeed input to dynamic_rnn
    encoder_output, encoder_state = tf.nn.dynamic_rnn(enc_cell, 
                                                     encoder_embedded_input, 
                                                     sequence_length=source_sequence_length,
                                                     dtype=tf.float32)
    
    return encoder_output, encoder_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    
     # Helper for the training process. Used by BasicDecoder to read inputs.
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                                            sequence_length=target_sequence_length,
                                                            time_major=False)
    # Basic decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, training_helper,encoder_state,output_layer) 
        
    # Perform dynamic decoding using the decoder
    training_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(training_decoder, impute_finished=True,
                                                                maximum_iterations=max_summary_length)
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [14]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """    
    
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')

    # Helper for the inference process.
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)

    # Basic decoder
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,inference_helper, encoder_state, output_layer)

    # Perform dynamic decoding using the decoder
    inference_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                        impute_finished=True,
                                                        maximum_iterations=max_target_sequence_length)
    
    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [21]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    
    # 1. Decoder Embedding
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
     
    # 3. Dense layer to translate the decoder's output at each time 
    # step into a choice from the target vocabulary
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))


    # 4. Set up a training decoder and an inference decoder
    # Training Decoder
    with tf.variable_scope("decode"):
        # Perform dynamic decoding using the decoder
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                                                       target_sequence_length, max_target_sequence_length, 
                                                       output_layer, keep_prob)
        
    # Reuses the same parameters trained by the training process
    with tf.variable_scope("decode", reuse=True):
        # Perform dynamic decoding using the decoder
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, 
                                                        target_vocab_to_int['<GO>'],target_vocab_to_int['<EOS>'], 
                                                        max_target_sequence_length, target_vocab_size, output_layer,
                                                        batch_size, keep_prob)

    return training_decoder_output, inference_decoder_output
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [27]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """

    # Pass the input data through the encoder. We'll ignore the encoder output, but use the state
    _, enc_state = encoding_layer(input_data, 
                                  rnn_size, 
                                  num_layers,
                                  keep_prob,
                                  source_sequence_length,
                                  source_vocab_size, 
                                  encoding_embedding_size=enc_embedding_size)
    
    
    # Prepare the target sequences we'll feed to the decoder in training mode
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    # Pass encoder state and decoder inputs to the decoders
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, enc_state, 
                                       target_sequence_length, max_target_sentence_length, 
                                       rnn_size, num_layers, target_vocab_to_int, target_vocab_size, 
                                       batch_size, keep_prob, dec_embedding_size)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [32]:
# Number of Epochs
epochs = 10
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 200
# Number of Layers
num_layers = 3
# Embedding Size
encoding_embedding_size = 200
decoding_embedding_size = 200
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.65
display_step = 2

### Build the Graph
Build the graph using the neural network you implemented.

In [33]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [30]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [34]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    2/269 - Train Accuracy: 0.0702, Validation Accuracy: 0.0654, Loss: 3.8154
Epoch   0 Batch    4/269 - Train Accuracy: 0.2774, Validation Accuracy: 0.3479, Loss: 3.6491
Epoch   0 Batch    6/269 - Train Accuracy: 0.3177, Validation Accuracy: 0.3469, Loss: 3.2429
Epoch   0 Batch    8/269 - Train Accuracy: 0.2847, Validation Accuracy: 0.3487, Loss: 3.2847
Epoch   0 Batch   10/269 - Train Accuracy: 0.3070, Validation Accuracy: 0.3738, Loss: 3.1421
Epoch   0 Batch   12/269 - Train Accuracy: 0.3230, Validation Accuracy: 0.3823, Loss: 3.1149
Epoch   0 Batch   14/269 - Train Accuracy: 0.3374, Validation Accuracy: 0.3756, Loss: 2.9254
Epoch   0 Batch   16/269 - Train Accuracy: 0.3564, Validation Accuracy: 0.3857, Loss: 2.8536
Epoch   0 Batch   18/269 - Train Accuracy: 0.3330, Validation Accuracy: 0.3991, Loss: 2.8940
Epoch   0 Batch   20/269 - Train Accuracy: 0.3365, Validation Accuracy: 0.4003, Loss: 2.9551
Epoch   0 Batch   22/269 - Train Accuracy: 0.3912, Validation Accuracy

Epoch   0 Batch  180/269 - Train Accuracy: 0.6197, Validation Accuracy: 0.6205, Loss: 0.7481
Epoch   0 Batch  182/269 - Train Accuracy: 0.6223, Validation Accuracy: 0.6309, Loss: 0.7593
Epoch   0 Batch  184/269 - Train Accuracy: 0.5994, Validation Accuracy: 0.6202, Loss: 0.7720
Epoch   0 Batch  186/269 - Train Accuracy: 0.6080, Validation Accuracy: 0.6357, Loss: 0.7505
Epoch   0 Batch  188/269 - Train Accuracy: 0.6169, Validation Accuracy: 0.6286, Loss: 0.7063
Epoch   0 Batch  190/269 - Train Accuracy: 0.6203, Validation Accuracy: 0.6316, Loss: 0.6932
Epoch   0 Batch  192/269 - Train Accuracy: 0.6231, Validation Accuracy: 0.6277, Loss: 0.7042
Epoch   0 Batch  194/269 - Train Accuracy: 0.6265, Validation Accuracy: 0.6317, Loss: 0.7032
Epoch   0 Batch  196/269 - Train Accuracy: 0.6319, Validation Accuracy: 0.6389, Loss: 0.6887
Epoch   0 Batch  198/269 - Train Accuracy: 0.6042, Validation Accuracy: 0.6350, Loss: 0.7246
Epoch   0 Batch  200/269 - Train Accuracy: 0.6263, Validation Accuracy

Epoch   1 Batch   92/269 - Train Accuracy: 0.7164, Validation Accuracy: 0.7031, Loss: 0.4115
Epoch   1 Batch   94/269 - Train Accuracy: 0.7239, Validation Accuracy: 0.7055, Loss: 0.4176
Epoch   1 Batch   96/269 - Train Accuracy: 0.7323, Validation Accuracy: 0.7188, Loss: 0.4052
Epoch   1 Batch   98/269 - Train Accuracy: 0.7358, Validation Accuracy: 0.7092, Loss: 0.4065
Epoch   1 Batch  100/269 - Train Accuracy: 0.7426, Validation Accuracy: 0.7011, Loss: 0.4017
Epoch   1 Batch  102/269 - Train Accuracy: 0.7229, Validation Accuracy: 0.7269, Loss: 0.3959
Epoch   1 Batch  104/269 - Train Accuracy: 0.7140, Validation Accuracy: 0.7186, Loss: 0.3911
Epoch   1 Batch  106/269 - Train Accuracy: 0.7239, Validation Accuracy: 0.7188, Loss: 0.3783
Epoch   1 Batch  108/269 - Train Accuracy: 0.7281, Validation Accuracy: 0.7243, Loss: 0.3875
Epoch   1 Batch  110/269 - Train Accuracy: 0.7424, Validation Accuracy: 0.7195, Loss: 0.3809
Epoch   1 Batch  112/269 - Train Accuracy: 0.7403, Validation Accuracy

Epoch   2 Batch    4/269 - Train Accuracy: 0.8249, Validation Accuracy: 0.8417, Loss: 0.2250
Epoch   2 Batch    6/269 - Train Accuracy: 0.8342, Validation Accuracy: 0.8493, Loss: 0.2020
Epoch   2 Batch    8/269 - Train Accuracy: 0.8389, Validation Accuracy: 0.8423, Loss: 0.2204
Epoch   2 Batch   10/269 - Train Accuracy: 0.8320, Validation Accuracy: 0.8476, Loss: 0.2114
Epoch   2 Batch   12/269 - Train Accuracy: 0.8371, Validation Accuracy: 0.8426, Loss: 0.2204
Epoch   2 Batch   14/269 - Train Accuracy: 0.8479, Validation Accuracy: 0.8455, Loss: 0.2064
Epoch   2 Batch   16/269 - Train Accuracy: 0.8355, Validation Accuracy: 0.8412, Loss: 0.2062
Epoch   2 Batch   18/269 - Train Accuracy: 0.8307, Validation Accuracy: 0.8410, Loss: 0.2068
Epoch   2 Batch   20/269 - Train Accuracy: 0.8513, Validation Accuracy: 0.8525, Loss: 0.2026
Epoch   2 Batch   22/269 - Train Accuracy: 0.8626, Validation Accuracy: 0.8480, Loss: 0.1879
Epoch   2 Batch   24/269 - Train Accuracy: 0.8415, Validation Accuracy

Epoch   2 Batch  182/269 - Train Accuracy: 0.8962, Validation Accuracy: 0.9073, Loss: 0.1015
Epoch   2 Batch  184/269 - Train Accuracy: 0.9156, Validation Accuracy: 0.8999, Loss: 0.0969
Epoch   2 Batch  186/269 - Train Accuracy: 0.9018, Validation Accuracy: 0.9050, Loss: 0.0919
Epoch   2 Batch  188/269 - Train Accuracy: 0.9110, Validation Accuracy: 0.9204, Loss: 0.0924
Epoch   2 Batch  190/269 - Train Accuracy: 0.9159, Validation Accuracy: 0.9093, Loss: 0.0925
Epoch   2 Batch  192/269 - Train Accuracy: 0.9088, Validation Accuracy: 0.9057, Loss: 0.0958
Epoch   2 Batch  194/269 - Train Accuracy: 0.8983, Validation Accuracy: 0.9038, Loss: 0.0955
Epoch   2 Batch  196/269 - Train Accuracy: 0.9001, Validation Accuracy: 0.9046, Loss: 0.0922
Epoch   2 Batch  198/269 - Train Accuracy: 0.9127, Validation Accuracy: 0.9050, Loss: 0.0990
Epoch   2 Batch  200/269 - Train Accuracy: 0.9177, Validation Accuracy: 0.8978, Loss: 0.0959
Epoch   2 Batch  202/269 - Train Accuracy: 0.9070, Validation Accuracy

Epoch   3 Batch   94/269 - Train Accuracy: 0.9290, Validation Accuracy: 0.9237, Loss: 0.0722
Epoch   3 Batch   96/269 - Train Accuracy: 0.9182, Validation Accuracy: 0.9268, Loss: 0.0661
Epoch   3 Batch   98/269 - Train Accuracy: 0.9336, Validation Accuracy: 0.9192, Loss: 0.0625
Epoch   3 Batch  100/269 - Train Accuracy: 0.9396, Validation Accuracy: 0.9267, Loss: 0.0633
Epoch   3 Batch  102/269 - Train Accuracy: 0.9357, Validation Accuracy: 0.9312, Loss: 0.0567
Epoch   3 Batch  104/269 - Train Accuracy: 0.9332, Validation Accuracy: 0.9292, Loss: 0.0571
Epoch   3 Batch  106/269 - Train Accuracy: 0.9429, Validation Accuracy: 0.9361, Loss: 0.0536
Epoch   3 Batch  108/269 - Train Accuracy: 0.9433, Validation Accuracy: 0.9341, Loss: 0.0568
Epoch   3 Batch  110/269 - Train Accuracy: 0.9282, Validation Accuracy: 0.9268, Loss: 0.0569
Epoch   3 Batch  112/269 - Train Accuracy: 0.9458, Validation Accuracy: 0.9249, Loss: 0.0641
Epoch   3 Batch  114/269 - Train Accuracy: 0.9221, Validation Accuracy

Epoch   4 Batch    6/269 - Train Accuracy: 0.9471, Validation Accuracy: 0.9379, Loss: 0.0439
Epoch   4 Batch    8/269 - Train Accuracy: 0.9348, Validation Accuracy: 0.9370, Loss: 0.0545
Epoch   4 Batch   10/269 - Train Accuracy: 0.9417, Validation Accuracy: 0.9395, Loss: 0.0434
Epoch   4 Batch   12/269 - Train Accuracy: 0.9255, Validation Accuracy: 0.9316, Loss: 0.0539
Epoch   4 Batch   14/269 - Train Accuracy: 0.9330, Validation Accuracy: 0.9336, Loss: 0.0456
Epoch   4 Batch   16/269 - Train Accuracy: 0.9426, Validation Accuracy: 0.9389, Loss: 0.0544
Epoch   4 Batch   18/269 - Train Accuracy: 0.9451, Validation Accuracy: 0.9382, Loss: 0.0440
Epoch   4 Batch   20/269 - Train Accuracy: 0.9416, Validation Accuracy: 0.9407, Loss: 0.0443
Epoch   4 Batch   22/269 - Train Accuracy: 0.9411, Validation Accuracy: 0.9331, Loss: 0.0422
Epoch   4 Batch   24/269 - Train Accuracy: 0.9347, Validation Accuracy: 0.9395, Loss: 0.0448
Epoch   4 Batch   26/269 - Train Accuracy: 0.9333, Validation Accuracy

Epoch   4 Batch  184/269 - Train Accuracy: 0.9655, Validation Accuracy: 0.9487, Loss: 0.0340
Epoch   4 Batch  186/269 - Train Accuracy: 0.9563, Validation Accuracy: 0.9446, Loss: 0.0317
Epoch   4 Batch  188/269 - Train Accuracy: 0.9549, Validation Accuracy: 0.9511, Loss: 0.0356
Epoch   4 Batch  190/269 - Train Accuracy: 0.9420, Validation Accuracy: 0.9471, Loss: 0.0411
Epoch   4 Batch  192/269 - Train Accuracy: 0.9558, Validation Accuracy: 0.9418, Loss: 0.0396
Epoch   4 Batch  194/269 - Train Accuracy: 0.9483, Validation Accuracy: 0.9458, Loss: 0.0376
Epoch   4 Batch  196/269 - Train Accuracy: 0.9516, Validation Accuracy: 0.9506, Loss: 0.0357
Epoch   4 Batch  198/269 - Train Accuracy: 0.9579, Validation Accuracy: 0.9494, Loss: 0.0397
Epoch   4 Batch  200/269 - Train Accuracy: 0.9612, Validation Accuracy: 0.9405, Loss: 0.0341
Epoch   4 Batch  202/269 - Train Accuracy: 0.9494, Validation Accuracy: 0.9470, Loss: 0.0376
Epoch   4 Batch  204/269 - Train Accuracy: 0.9500, Validation Accuracy

Epoch   5 Batch   96/269 - Train Accuracy: 0.9479, Validation Accuracy: 0.9553, Loss: 0.0371
Epoch   5 Batch   98/269 - Train Accuracy: 0.9467, Validation Accuracy: 0.9474, Loss: 0.0380
Epoch   5 Batch  100/269 - Train Accuracy: 0.9567, Validation Accuracy: 0.9512, Loss: 0.0375
Epoch   5 Batch  102/269 - Train Accuracy: 0.9481, Validation Accuracy: 0.9492, Loss: 0.0346
Epoch   5 Batch  104/269 - Train Accuracy: 0.9654, Validation Accuracy: 0.9554, Loss: 0.0333
Epoch   5 Batch  106/269 - Train Accuracy: 0.9708, Validation Accuracy: 0.9573, Loss: 0.0297
Epoch   5 Batch  108/269 - Train Accuracy: 0.9706, Validation Accuracy: 0.9537, Loss: 0.0304
Epoch   5 Batch  110/269 - Train Accuracy: 0.9582, Validation Accuracy: 0.9548, Loss: 0.0325
Epoch   5 Batch  112/269 - Train Accuracy: 0.9559, Validation Accuracy: 0.9523, Loss: 0.0383
Epoch   5 Batch  114/269 - Train Accuracy: 0.9605, Validation Accuracy: 0.9473, Loss: 0.0326
Epoch   5 Batch  116/269 - Train Accuracy: 0.9656, Validation Accuracy

Epoch   6 Batch    8/269 - Train Accuracy: 0.9632, Validation Accuracy: 0.9585, Loss: 0.0320
Epoch   6 Batch   10/269 - Train Accuracy: 0.9667, Validation Accuracy: 0.9612, Loss: 0.0253
Epoch   6 Batch   12/269 - Train Accuracy: 0.9533, Validation Accuracy: 0.9600, Loss: 0.0354
Epoch   6 Batch   14/269 - Train Accuracy: 0.9592, Validation Accuracy: 0.9548, Loss: 0.0276
Epoch   6 Batch   16/269 - Train Accuracy: 0.9564, Validation Accuracy: 0.9535, Loss: 0.0360
Epoch   6 Batch   18/269 - Train Accuracy: 0.9639, Validation Accuracy: 0.9550, Loss: 0.0307
Epoch   6 Batch   20/269 - Train Accuracy: 0.9609, Validation Accuracy: 0.9602, Loss: 0.0281
Epoch   6 Batch   22/269 - Train Accuracy: 0.9627, Validation Accuracy: 0.9569, Loss: 0.0291
Epoch   6 Batch   24/269 - Train Accuracy: 0.9655, Validation Accuracy: 0.9594, Loss: 0.0266
Epoch   6 Batch   26/269 - Train Accuracy: 0.9685, Validation Accuracy: 0.9573, Loss: 0.0270
Epoch   6 Batch   28/269 - Train Accuracy: 0.9527, Validation Accuracy

Epoch   6 Batch  186/269 - Train Accuracy: 0.9633, Validation Accuracy: 0.9604, Loss: 0.0219
Epoch   6 Batch  188/269 - Train Accuracy: 0.9742, Validation Accuracy: 0.9657, Loss: 0.0255
Epoch   6 Batch  190/269 - Train Accuracy: 0.9590, Validation Accuracy: 0.9536, Loss: 0.0294
Epoch   6 Batch  192/269 - Train Accuracy: 0.9734, Validation Accuracy: 0.9561, Loss: 0.0296
Epoch   6 Batch  194/269 - Train Accuracy: 0.9626, Validation Accuracy: 0.9513, Loss: 0.0267
Epoch   6 Batch  196/269 - Train Accuracy: 0.9526, Validation Accuracy: 0.9471, Loss: 0.0278
Epoch   6 Batch  198/269 - Train Accuracy: 0.9602, Validation Accuracy: 0.9596, Loss: 0.0279
Epoch   6 Batch  200/269 - Train Accuracy: 0.9787, Validation Accuracy: 0.9603, Loss: 0.0235
Epoch   6 Batch  202/269 - Train Accuracy: 0.9576, Validation Accuracy: 0.9538, Loss: 0.0309
Epoch   6 Batch  204/269 - Train Accuracy: 0.9623, Validation Accuracy: 0.9557, Loss: 0.0261
Epoch   6 Batch  206/269 - Train Accuracy: 0.9544, Validation Accuracy

Epoch   7 Batch   98/269 - Train Accuracy: 0.9622, Validation Accuracy: 0.9619, Loss: 0.0256
Epoch   7 Batch  100/269 - Train Accuracy: 0.9651, Validation Accuracy: 0.9646, Loss: 0.0304
Epoch   7 Batch  102/269 - Train Accuracy: 0.9600, Validation Accuracy: 0.9660, Loss: 0.0244
Epoch   7 Batch  104/269 - Train Accuracy: 0.9669, Validation Accuracy: 0.9600, Loss: 0.0260
Epoch   7 Batch  106/269 - Train Accuracy: 0.9769, Validation Accuracy: 0.9671, Loss: 0.0219
Epoch   7 Batch  108/269 - Train Accuracy: 0.9710, Validation Accuracy: 0.9632, Loss: 0.0217
Epoch   7 Batch  110/269 - Train Accuracy: 0.9665, Validation Accuracy: 0.9573, Loss: 0.0240
Epoch   7 Batch  112/269 - Train Accuracy: 0.9657, Validation Accuracy: 0.9569, Loss: 0.0273
Epoch   7 Batch  114/269 - Train Accuracy: 0.9654, Validation Accuracy: 0.9581, Loss: 0.0237
Epoch   7 Batch  116/269 - Train Accuracy: 0.9763, Validation Accuracy: 0.9606, Loss: 0.0248
Epoch   7 Batch  118/269 - Train Accuracy: 0.9657, Validation Accuracy

Epoch   8 Batch   10/269 - Train Accuracy: 0.9736, Validation Accuracy: 0.9661, Loss: 0.0216
Epoch   8 Batch   12/269 - Train Accuracy: 0.9500, Validation Accuracy: 0.9624, Loss: 0.0265
Epoch   8 Batch   14/269 - Train Accuracy: 0.9639, Validation Accuracy: 0.9645, Loss: 0.0233
Epoch   8 Batch   16/269 - Train Accuracy: 0.9593, Validation Accuracy: 0.9599, Loss: 0.0296
Epoch   8 Batch   18/269 - Train Accuracy: 0.9694, Validation Accuracy: 0.9548, Loss: 0.0254
Epoch   8 Batch   20/269 - Train Accuracy: 0.9645, Validation Accuracy: 0.9595, Loss: 0.0243
Epoch   8 Batch   22/269 - Train Accuracy: 0.9631, Validation Accuracy: 0.9595, Loss: 0.0250
Epoch   8 Batch   24/269 - Train Accuracy: 0.9739, Validation Accuracy: 0.9608, Loss: 0.0233
Epoch   8 Batch   26/269 - Train Accuracy: 0.9643, Validation Accuracy: 0.9588, Loss: 0.0212
Epoch   8 Batch   28/269 - Train Accuracy: 0.9588, Validation Accuracy: 0.9625, Loss: 0.0235
Epoch   8 Batch   30/269 - Train Accuracy: 0.9643, Validation Accuracy

Epoch   8 Batch  188/269 - Train Accuracy: 0.9775, Validation Accuracy: 0.9632, Loss: 0.0241
Epoch   8 Batch  190/269 - Train Accuracy: 0.9702, Validation Accuracy: 0.9594, Loss: 0.0241
Epoch   8 Batch  192/269 - Train Accuracy: 0.9750, Validation Accuracy: 0.9550, Loss: 0.0247
Epoch   8 Batch  194/269 - Train Accuracy: 0.9616, Validation Accuracy: 0.9504, Loss: 0.0246
Epoch   8 Batch  196/269 - Train Accuracy: 0.9594, Validation Accuracy: 0.9539, Loss: 0.0276
Epoch   8 Batch  198/269 - Train Accuracy: 0.9588, Validation Accuracy: 0.9546, Loss: 0.0271
Epoch   8 Batch  200/269 - Train Accuracy: 0.9713, Validation Accuracy: 0.9600, Loss: 0.0236
Epoch   8 Batch  202/269 - Train Accuracy: 0.9652, Validation Accuracy: 0.9578, Loss: 0.0288
Epoch   8 Batch  204/269 - Train Accuracy: 0.9631, Validation Accuracy: 0.9681, Loss: 0.0231
Epoch   8 Batch  206/269 - Train Accuracy: 0.9556, Validation Accuracy: 0.9680, Loss: 0.0286
Epoch   8 Batch  208/269 - Train Accuracy: 0.9714, Validation Accuracy

Epoch   9 Batch  100/269 - Train Accuracy: 0.9688, Validation Accuracy: 0.9660, Loss: 0.0237
Epoch   9 Batch  102/269 - Train Accuracy: 0.9681, Validation Accuracy: 0.9646, Loss: 0.0197
Epoch   9 Batch  104/269 - Train Accuracy: 0.9739, Validation Accuracy: 0.9594, Loss: 0.0203
Epoch   9 Batch  106/269 - Train Accuracy: 0.9785, Validation Accuracy: 0.9717, Loss: 0.0191
Epoch   9 Batch  108/269 - Train Accuracy: 0.9734, Validation Accuracy: 0.9727, Loss: 0.0173
Epoch   9 Batch  110/269 - Train Accuracy: 0.9588, Validation Accuracy: 0.9638, Loss: 0.0203
Epoch   9 Batch  112/269 - Train Accuracy: 0.9620, Validation Accuracy: 0.9603, Loss: 0.0222
Epoch   9 Batch  114/269 - Train Accuracy: 0.9687, Validation Accuracy: 0.9615, Loss: 0.0205
Epoch   9 Batch  116/269 - Train Accuracy: 0.9691, Validation Accuracy: 0.9584, Loss: 0.0207
Epoch   9 Batch  118/269 - Train Accuracy: 0.9594, Validation Accuracy: 0.9665, Loss: 0.0188
Epoch   9 Batch  120/269 - Train Accuracy: 0.9595, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [35]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [36]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [37]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    preprocced_list=[]
    for word in sentence.split():
        preprocced_list.append(vocab_to_int.get(word.lower(), vocab_to_int.get('<UNK>')))
    
    return preprocced_list

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [41]:
translate_sentence = 'I love you .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [157, 2, 192, 30]
  English Words: ['i', '<UNK>', 'you', '.']

Prediction
  Word Ids:      [107, 214, 19, 276, 1]
  French Words: je est février . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.